### PR TITLE
[Experimental] Wrap around skaffold dev

### PR DESCRIPTION
**Pull request by marcinc**
Addresses #293 

**NOTE: This feature is EXPERIMENTAL**

This PR:

- [x] Adds option to activate `skaffold dev` loop when iterating on the project with `kev dev` with `--skaffold` flag.
- [x] Ensure that `skaffold dev` is properly configured before starting the loop:
  - [x] the right skaffold profile(s) are used
  - [x] the correct target k8s namespace is passed
  - [x] the correct kubecontext is referenced pointing skaffold at the correct cluster

Usage:

   To start iterating on the application with Kev and automatically activate skaffold dev loop, pass `--skaffold` flag to `kev dev` command. 

Skaffold will expect the following options to work:
* `--kubecontext | -k` - name of the kubecontext to be used by skaffold. This will point it to the correct K8s cluster. Defaults to current kubectl context.
* `--namespace | -n` - name of target K8s namespace to which automatic deployment should happen. Defaults to `default` if not specified.
* `--kev-env` - name of one of Kev managed environments. This will tell Skaffold which manifests to deploy. Defaults to first of specified environments (via `--environments | -e`) or default `dev` environment if none have be provided.

Breaking the dev loop with Ctrl+C will automatically cleanup deployment target namespace and remove all objects. It'll also attempt to prune locally built docker images. Be mindful, that prune operation may take a while and could fail so user must keep an eye on number and size of images built locally. Auto pruning before dev loop will be added in the future Skaffold versions and should help keep that in check.

### CODE REVIEW

File: cmd/kev/cmd/dev.go [hunk 1]
```diff
@@ -167,9 +167,21 @@ func runDevCmd(cmd *cobra.Command, args []string) error {
 		return err
```
Comment (ezodude) [inline cmd/kev/cmd/dev.go:168]: To keep err handling consistent with other cmds, just use `return displayError(err)`.

Should prob apply to all `return err` in this file.

File: cmd/kev/cmd/dev.go [hunk 2]
```diff
@@ -167,9 +167,21 @@ func runDevCmd(cmd *cobra.Command, args []string) error {
...
+		return err
+	}
+
 	if skaffold {
-		profileName := kevenv + kev.EnvProfileNameSuffix
-		go kev.RunSkaffoldDev(cmd.Context(), cmd.OutOrStdout(), []string{profileName}, namespace, kubecontext, 1000)
+		if skaffoldConfigPath, skaffoldConfig, ok := kev.ActivateSkaffoldDevLoop(workDir); ok {
+			if err := writeTo(skaffoldConfigPath, skaffoldConfig); err != nil {
+				log.Warnf("Couldn't write Skaffold config: %s.", err)
+				return err
```
Comment (ezodude) [inline cmd/kev/cmd/dev.go:180]: How about if we simply handle this as a log.Error (and use `return displayError(err)`)?

File: pkg/kev/kev.go [hunk 3]
```diff
@@ -183,3 +184,41 @@ func Watch(workDir string, envs []string, change chan<- string) error {
...
+		log.Warn(`Can't activate Skaffold dev loop. Kev wasn't initialized with --skaffold.
+	If you don't currently have skaffold.yaml in your project you may bootstrap a new one with "skaffold init" command.
+	Once you have skaffold.yaml in your project, make sure that Kev references it by adding "skaffold: skaffold.yaml" in kev.yaml!`)
+		return "", nil, false
+	}
+
+	configPath := path.Join(workDir, manifest.Skaffold)
+
+	if !fileExists(configPath) {
+		log.Error(`Can't file Skaffold config file referenced by Kev manifest. Have you initialized Kev with --skaffold?
```
Comment (ezodude) [inline pkg/kev/kev.go:210]: `log.Error(`Can't file Skaffold` -> `log.Error(`Can't find Skaffold`

**Review by ezodude — COMMENTED** [review]
Make sense.

Generally, feels like we might need to start thinking about a top level Plugins package. As `Skaffold` logic is starting to gently bleed into core `Kev` logic.

File: pkg/kev/skaffold.go [hunk 4]
```diff
@@ -452,3 +463,116 @@ func (s *SkaffoldManifest) sortProfiles() {
...
+	runCtx, cfg, err := runContext(opts)
+
+	r, err := runner.NewForConfig(runCtx)
+
+	if err != nil {
+		log.Error("Skaffold dev failed", err)
+		return err
+	}
+
+	prune := func() {}
```
[the remark this answers is not in the document]
Comment (marcinc) [reply pkg/kev/skaffold.go:506]: Yes it's a bit confusing at first. These bits have been extracted from skaffold codebase, mind. From what I gathered it was to ensure the ordering of Pruning and Cleanup i.e. to ensure Pruning happens after Cleanup. I've now updated the code to actually make use of those when Ctrl+C is intercepted.

File: cmd/kev/cmd/dev.go (hunk 1 above)
Comment (marcinc) [reply cmd/kev/cmd/dev.go]: updated

File: cmd/kev/cmd/dev.go (hunk 2 above)
Comment (marcinc) [reply cmd/kev/cmd/dev.go]: updated

**Review by ezodude — APPROVED** [review]
👍

**Pull request merged**